### ISSUE TITLE
[Breaking change]: Regexes that has ranges in the pattern and used with RegexOptions.IgnoreCase option might break

### ISSUE THREAD

**Issue by buyaa-n** [dotnet/docs#36733]
### Description

[Regex was incorrectly handling casing of some ranges](https://github.com/dotnet/runtime/issues/36149) in .netframework, .NET 6 and below. The two patterns in below example should produce the same result, because the range from `\xD7` through `\xD8` is just same as `\xD7` and `\xD8`, since there's nothing in between them. 

The fix for this bug could cause a breaking change in case your Regex had a bug that is not revealed because of this bug or if you had a workaround for this bug. See a broken Regex example from https://github.com/dotnet/runtime/issues/78609

### Version

.NET 7

### Previous behavior

The first match correctly produces `false` whereas the second incorrectly produces `true`:
```cs
Regex.IsMatch("\xF7", @"^(?i:[\xD7\xD8])$", RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
Regex.IsMatch("\xF7", @"^(?i:[\xD7-\xD8])$", RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
```

### New behavior

As the bug fixed in .NET 7, now both cases produce `false` correctly.

### Type of breaking change

- [ ] **Binary incompatible**: Existing binaries may encounter a breaking change in behavior, such as failure to load or execute, and if so, require recompilation.
- [ ] **Source incompatible**: When recompiled using the new SDK or component or to target the new runtime, existing source code may require source changes to compile successfully.
- [X] **Behavioral change**: Existing binaries may behave differently at run time.

### Reason for change

The previous behavior was incorrect, as the Regexes in the example should produce the same result

### Recommended action

If you Regex has a hidden bug, fix it, if you had a workaround for this bug, you could remove the workaround.

### Feature area

Core .NET libraries

### Affected APIs

The change is introduced in the Regex range handling, not for any specific API

---
[Associated WorkItem - 189691](https://dev.azure.com/msft-skilling/Content/_workitems/edit/189691)

**Comment by buyaa-n** [comment]
CC @stephentoub

**Comment by stephentoub** [comment]
@ericstj, is there a particular reason this bug fix deserves a breaking change document?

**Comment by ericstj** [comment]
I think the reasoning was discussed here: https://github.com/dotnet/runtime/issues/78609#issuecomment-1346922775.  Customer was broken by the change, we decided it was by design and a good fix. To help others who might encounter it in the future we can document it as breaking and intentional.  The goal here would be to leave enough of a trail so that the next person might see that and not open an issue but instead find a solution.

Note: I unchecked binary incompatible - this is not a binary breaking change, it's only a functional difference.
